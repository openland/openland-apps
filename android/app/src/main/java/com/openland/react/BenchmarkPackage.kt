package com.openland.react

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class BenchmarkModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "BenchmarkModule"
    }

    @ReactMethod
    fun setConfig() {
    }
}