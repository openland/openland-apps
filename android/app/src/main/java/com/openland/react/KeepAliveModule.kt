package com.openland.react

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.jstasks.HeadlessJsTaskContext
//import com.facebook.react.modules.core.Timing
import java.util.concurrent.atomic.AtomicBoolean

class KeepAliveModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String {
        return "RNKeepAlive"
    }

    override fun initialize() {
        
        // Hack Timers
        runOnUIThread {
            val headlessContext = HeadlessJsTaskContext.getInstance(reactApplicationContext)
            try {
                val mActiveTasks = HeadlessJsTaskContext::class.java.getDeclaredField("mActiveTasks")
                mActiveTasks.isAccessible = true
                val tasks = mActiveTasks.get(headlessContext) as MutableSet<Int>
                tasks.add(-1)
            } catch (e: NoSuchFieldException) {
                e.printStackTrace()
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            }

//            val module = reactApplicationContext.getNativeModule(Timing::class.java)
//            try {
//                val isRunningTasks = Timing::class.java.getDeclaredField("isRunningTasks")
//                isRunningTasks.isAccessible = true
//                val v = isRunningTasks.get(module) as AtomicBoolean
//                v.set(true)
//            } catch (e: NoSuchFieldException) {
//                e.printStackTrace()
//            } catch (e: IllegalAccessException) {
//                e.printStackTrace()
//            }
        }
    }
}