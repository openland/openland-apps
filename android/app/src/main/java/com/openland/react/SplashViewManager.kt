package com.openland.react

import android.content.res.Configuration
import android.graphics.drawable.AnimatedVectorDrawable
import android.util.Log
import android.view.View
import com.facebook.react.uimanager.ThemedReactContext
import android.view.LayoutInflater
import android.widget.ImageView
import androidx.appcompat.content.res.AppCompatResources.getDrawable
import androidx.vectordrawable.graphics.drawable.AnimatedVectorDrawableCompat
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ViewProps
import com.facebook.react.uimanager.annotations.ReactProp
import com.openland.app.R

class SplashViewManager : SimpleViewManager<View>() {

    override fun getName(): String {
        return "AndroidSplashView"
    }

    override fun createViewInstance(reactContext: ThemedReactContext): View {
        val activity = if (reactContext.currentActivity != null) reactContext.currentActivity!! else reactContext
        val currentMode = activity.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
        val isDark = currentMode == Configuration.UI_MODE_NIGHT_YES
        val inflater = LayoutInflater.from(activity)
        val view = inflater.inflate(if (isDark) R.layout.splash_dark else R.layout.splash, null)
        return view
    }


    @ReactProp(name = "splashVisible")
    fun setSplashVisible(view: View, visible: Boolean) {
        if (!visible && view.getTag(R.id.visible) != true) {
            view.setTag(R.id.visible, true)
            for(i in 1..10){
                view.postDelayed({
                    Thread.sleep(500)
                }, 500)
            }
            view.animate().alpha(0f).setDuration(250L).start()
//            (view.findViewById<ImageView>(R.id.image).background as AnimatedVectorDrawable).start()
//            (view.findViewById<ImageView>(R.id.back).background as AnimatedVectorDrawable).start()
        }
    }

}
