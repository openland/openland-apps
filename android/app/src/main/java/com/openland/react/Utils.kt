package com.openland.react

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.jstasks.HeadlessJsTaskContext
import com.facebook.react.jstasks.HeadlessJsTaskEventListener

private val handler = Handler(Looper.getMainLooper())

fun runOnUIThreadDelayed(delay: Int, callback: () -> Unit) {
    handler.postDelayed(callback, delay.toLong())
}

fun runOnUIThread(callback: () -> Unit) {
    handler.post(callback)
}

val Context.statusBarHeight: Float
    get() {
        val heightResId = this.resources
                .getIdentifier("status_bar_height", "dimen", "android")

        val statusBarHeight = if (heightResId > 0)
            this.resources.getDimensionPixelSize(heightResId).toFloat()
        else
            0.0f

        return statusBarHeight
    }


fun watchLifecycle(reactContext: ReactApplicationContext, callback: (active: Boolean) -> Unit) {

    val headlessJsTaskContext = HeadlessJsTaskContext.getInstance(reactContext)

    val syncObject = Object()
    var lastValue = false

    var wasStarted = false
    var hasTask = headlessJsTaskContext.hasActiveTasks()

    fun notifyIfNeeded() {
        val nv = wasStarted or hasTask
        if (nv != lastValue) {
            lastValue = nv
            callback(nv)
        }
    }

    reactContext.addLifecycleEventListener(object : LifecycleEventListener {
        override fun onHostResume() {
            synchronized(syncObject) {
                wasStarted = true
                notifyIfNeeded()
            }
        }

        override fun onHostPause() {
            synchronized(syncObject) {
                wasStarted = false
                notifyIfNeeded()
            }
        }

        override fun onHostDestroy() {
            synchronized(syncObject) {
                wasStarted = false
                notifyIfNeeded()
            }
        }
    })

    headlessJsTaskContext.addTaskEventListener(object : HeadlessJsTaskEventListener {

        override fun onHeadlessJsTaskStart(taskId: Int) {
            synchronized(syncObject) {
                hasTask = headlessJsTaskContext.hasActiveTasks()
                notifyIfNeeded()
            }
        }

        override fun onHeadlessJsTaskFinish(taskId: Int) {
            synchronized(syncObject) {
                hasTask = headlessJsTaskContext.hasActiveTasks()
                notifyIfNeeded()
            }
        }
    })
}