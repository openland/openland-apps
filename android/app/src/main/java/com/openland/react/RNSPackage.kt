package com.openland.react

import android.view.View
import com.facebook.litho.config.ComponentsConfiguration
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ReactShadowNode
import com.facebook.react.uimanager.ViewManager
import com.openland.react.async.AsyncConfigManager
import com.openland.react.async.AsyncDataViewManager
import com.openland.react.async.AsyncListViewManager
import com.openland.react.async.AsyncViewManager
import com.openland.react.anim.RNSAnimatedViewManager
import com.openland.react.anim.RNSAnimatedViewViewManager
import com.openland.react.graphql.RNGraphQL
import com.openland.react.window.BottomSafeAreaProvider
import com.openland.react.window.RNSWindowManager

class RNSPackage : ReactPackage {

    init {
        ComponentsConfiguration.incrementalMountWhenNotVisible = false
    }

    override fun createNativeModules(reactContext: ReactApplicationContext): MutableList<NativeModule> {
        return mutableListOf(
                AsyncConfigManager(reactContext),
                AsyncDataViewManager(reactContext),
                RNSAnimatedViewManager(reactContext),
                RNSWindowManager(reactContext),
                BottomSafeAreaProvider(reactContext),
                BenchmarkModule(reactContext),
                RNGraphQL(reactContext)
        )
    }

    override fun createViewManagers(reactContext: ReactApplicationContext): MutableList<ViewManager<View, ReactShadowNode<*>>> {
        return mutableListOf(
                AsyncViewManager(),
                AsyncListViewManager(),
                RNSAnimatedViewViewManager.sharedInstance
        ) as MutableList<ViewManager<View, ReactShadowNode<*>>>
    }
}