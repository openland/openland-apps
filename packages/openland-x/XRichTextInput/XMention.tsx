import * as React from 'react';
import { MessageFull_mentions } from 'openland-api/Types';
import { getEmojiRegex } from 'openland-y-utils/getEmojiRegex';
import { UserPopper } from 'openland-web/components/messenger/components/view/content/UserPopper';

export const removeEmojiFromText = (str: string) => {
    return str ? str.replace(getEmojiRegex(), '').trim() : '';
};

const getMentionString = (str: string) => {
    return `@${removeEmojiFromText(str)}`;
};

type MentionComponentInnerTextProps = {
    className?: string;
    mention: MessageFull_mentions;
    hasPopper?: boolean;
    children?: any;
    inCompose?: boolean;
};

class MentionComponentInnerText extends React.PureComponent<{
    mention: MessageFull_mentions;
    inCompose?: boolean;
    children?: any;
}> {
    render() {
        const { mention, inCompose, children } = this.props;
        const paddings = inCompose
            ? {
                  paddingTop: 1,
                  paddingBottom: 1,
                  paddingLeft: 3,
                  paddingRight: 3,
              }
            : {};

        let style;

        if (mention.isYou) {
            style = {};
        }

        if (mention.isYou) {
            style = {
                ...paddings,
                backgroundColor: '#fff6e5',
                color: '#1790ff',
            };
        } else {
            style = {
                ...paddings,
                color: '#1790ff',
            };
        }

        return <span style={style}>{children}</span>;
    }
}

export class MentionComponentInner extends React.PureComponent<
    MentionComponentInnerTextProps
> {
    render() {
        const { hasPopper, mention } = this.props;
        if (hasPopper && mention) {
            return (
                <UserPopper
                    user={mention}
                    isMe={mention.isYou}
                    noCardOnMe
                    startSelected={false}
                >
                    <MentionComponentInnerText {...this.props} />
                </UserPopper>
            );
        } else {
            return <MentionComponentInnerText {...this.props} />;
        }
    }
}

type textBeforeMentionAndMentionT = {
    textBeforeMention: string;
    mention: MessageFull_mentions | null;
};

const getArrayOfPairsTextBeforeMentionAndMention = ({
    messageText,
    mentions,
}: {
    messageText: string;
    mentions: MessageFull_mentions[];
}): textBeforeMentionAndMentionT[] => {
    let mentionMatchesMap: { [key: number]: string } = {};
    let endText = messageText;

    let result = [];
    for (let i = 0; i < mentions.length; i++) {
        const mention = mentions[i];
        const mentionString = getMentionString(mention.name);
        const mentionIndex = endText.indexOf(mentionString);
        mentionMatchesMap[mentionIndex] = mentionString;
        const begining = endText.slice(0, mentionIndex);
        const end = endText.slice(mentionIndex + mentionString.length);

        result.push({
            textBeforeMention: begining,
            mention,
        });

        endText = end;
    }

    result.push({
        textBeforeMention: endText,
        mention: null,
    });

    return result;
};

export class MessageWithMentionsTextComponent extends React.PureComponent<{
    text: string;
    mentions: MessageFull_mentions[];
}> {
    render() {
        const { text, mentions } = this.props;

        return (
            <>
                {getArrayOfPairsTextBeforeMentionAndMention({
                    messageText: text,
                    mentions,
                }).map(
                    (
                        {
                            textBeforeMention,
                            mention,
                        }: textBeforeMentionAndMentionT,
                        key: number,
                    ) => {
                        return (
                            <span key={key}>
                                {textBeforeMention}
                                {mention && (
                                    <MentionComponentInner
                                        mention={mention}
                                        hasPopper
                                    >
                                        {removeEmojiFromText(mention.name)}
                                    </MentionComponentInner>
                                )}
                            </span>
                        );
                    },
                )}
            </>
        );
    }
}
