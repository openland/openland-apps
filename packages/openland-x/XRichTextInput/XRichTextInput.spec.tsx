import * as React from 'react';
import { convertToHTML, convertFromHTML } from 'draft-convert';
import { MessageWithMentionsTextComponent } from './XMention';
import * as cheerio from 'cheerio';
import * as renderer from 'react-test-renderer';

export const toContentState = (html: any, options?: any) => {
    return convertFromHTML({
        htmlToEntity: (nodeName: any, node: any, createEntity: any) => {
            if (nodeName === 'span') {
                return createEntity('MENTION', 'IMMUTABLE', {
                    'data-mention-id': node.getAttribute('data-mention-id'),
                });
            }
        },
    })(html, options);
};

export const toHTML = (contentState: any) => {
    return convertToHTML({
        entityToHTML: (entity: any, originalText: any) => {
            if (entity.type === 'MENTION') {
                return `<span data-mention-id="${
                    entity.data['data-mention-id']
                }">${originalText}</span>`;
            }
            return originalText;
        },
    })(contentState);
};

const testFromHtmlToDraftEditorStateAndBack = (htmlFixture: any) => {
    const contentState = toContentState(htmlFixture);
    const htmlOut = toHTML(contentState);
    expect(htmlOut).toBe(htmlFixture);
};

const htmlMessageToDbFormat = (str: string) => {
    const $ = cheerio.load(str);
    const mentions: any = [];

    $('span').each((index, item) => {
        mentions.push(item.attribs['data-mention-id']);
    });

    return {
        text: ($ as any).text(),
        mentions,
    };
};

describe.only('XRichComponent', () => {
    it('converts one mention with span', () => {
        testFromHtmlToDraftEditorStateAndBack(
            '<p><span data-mention-id="123">@Sergey Lapin</span></p>',
        );
    });

    it('converts two mention with span', () => {
        testFromHtmlToDraftEditorStateAndBack(
            '<p><span data-mention-id="123">@Sergey Lapin</span> <span data-mention-id="456">@Other Guy</span></p>',
        );
    });

    it('html to db format', () => {
        const str =
            '<p><span data-mention-id="123">@Sergey Lapin</span> 123 <span data-mention-id="456">@Other Guy</span></p>';

        const res = htmlMessageToDbFormat(str);

        expect(res).toEqual({
            mentions: ['123', '456'],
            text: '@Sergey Lapin 123 @Other Guy',
        });
    });

    it('db format to react component', () => {
        expect(
            renderer
                .create(
                    <MessageWithMentionsTextComponent
                        text={'@Sergey Lapin 123 @Other Guy'}
                        mentions={[
                            { id: '123', name: '@Sergey Lapin' } as any,
                            { id: '456', name: '@Other Guy' } as any,
                        ]}
                    />,
                )
                .toJSON(),
        ).toMatchSnapshot();
    });
});
