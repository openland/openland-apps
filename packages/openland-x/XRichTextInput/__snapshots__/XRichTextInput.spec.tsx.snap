// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XRichComponent db format to react component 1`] = `
<span>
  @Sergey Lapin 123 @Other Guy
  <span
    style={
      Object {
        "color": "#1790ff",
      }
    }
  >
    @Sergey Lapin
  </span>
</span>
`;
